File: python/agentic-ai/01-foundations/modern-llm-primitives/structured-output.py
from ollama import chat
from pydantic import BaseModel

MODEL = "qwen3.5:9b"

# -----------------------------
# Schema
# -----------------------------
class ResearchResult(BaseModel):
    topic: str
    summary: str
    sources: list[str]
    confidence: float
    needs_more_research: bool

# -----------------------------
# Run query
# -----------------------------
response = chat(
    model=MODEL,
    messages=[
        {"role": "system", "content": "You are a research assistant."},
        {"role": "user", "content": "Research the current state of MCP (Model Context Protocol)."},
    ],
    format=ResearchResult.model_json_schema(),
    options={"temperature": 0.2},
)

# -----------------------------
# Parse response
# -----------------------------
result = ResearchResult.model_validate_json(response.message.content)

print(f"Topic: {result.topic}")
print(f"Summary: {result.summary}")
print(f"Confidence: {result.confidence:.0%}")
print(f"Needs more research: {result.needs_more_research}")
print(f"Sources: {result.sources}")

File: rust/agentic-ai/r-01-d-modern-llm-primitives/src/structured_output.rs
use ollama_rs::{
    generation::chat::{ChatMessage, request::ChatMessageRequest},
    Ollama,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize)]
struct ResearchResult {
    topic: String,
    summary: String,
    sources: Vec<String>,
    confidence: f32,
    needs_more_research: bool,
}

// -----------------------------
// System Prompt (force JSON)
// -----------------------------
const SYSTEM_PROMPT: &str = r#"
You are a research assistant.

Return ONLY valid JSON matching this schema:

{
  "topic": string,
  "summary": string,
  "sources": string[],
  "confidence": number (0 to 1),
  "needs_more_research": boolean
}

Do not include any explanation or extra text.
"#;

pub async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let ollama = Ollama::default();

    let messages = vec![
        ChatMessage::system(SYSTEM_PROMPT.to_string()),
        ChatMessage::user("Research the current state of MCP (Model Context Protocol).".to_string()),
    ];

    let request = ChatMessageRequest::new("qwen3.5:9b".to_string(), messages);

    let response = ollama.send_chat_messages(request).await?;

    let content = response.message.content;

    // -----------------------------
    // Parse JSON → serde
    // -----------------------------
    let result: ResearchResult = serde_json::from_str(&content)?;

    println!("Topic: {}", result.topic);
    println!("Summary: {}", result.summary);
    println!("Confidence: {:.0}%", result.confidence * 100.0);
    println!("Needs more research: {}", result.needs_more_research);
    println!("Sources: {:?}", result.sources);

    Ok(())
}